tasks/base.py

```diff
@@ -375,3 +390,51 @@ def on_failure(self, exc, task_id, args, kwargs, einfo):
             TestResultsFlow.log(TestResultsFlow.CELERY_FAILURE)
 
         return res
+
+    def get_repo_provider_service(
+        self,
+        repository: Repository,
+        installation_name_to_use: str = GITHUB_APP_INSTALLATION_DEFAULT_NAME,
+        additional_data: AdditionalData = None,
+        commit: Commit = None,
+    ) -> TorngitBaseAdapter | None:
+        try:
+            return get_repo_provider_service(
+                repository, installation_name_to_use, additional_data
+            )
+        except RepositoryWithoutValidBotError:
+            save_commit_error(
+                commit,
+                error_code=CommitErrorTypes.REPO_BOT_INVALID.value,
+                error_params=dict(repoid=repository.repoid),
+            )
+            log.warning(
+                "Unable to reach git provider because repo doesn't have a valid bot"
+            )
+        except NoConfiguredAppsAvailable as exp:
+            if exp.rate_limited_count > 0:
+                # There's at least 1 app that we can use to communicate with GitHub,
+                # but this app happens to be rate limited now. We try again later.
+                # Min wait time of 1 minute
+                retry_delay_seconds = max(60, get_seconds_to_next_hour())
+                log.warning(
+                    "Unable to get repo provider service due to rate limits. Retrying again later.",
+                    extra=dict(
+                        apps_available=exp.apps_count,
+                        apps_rate_limited=exp.rate_limited_count,
+                        apps_suspended=exp.suspended_count,
+                        countdown_seconds=retry_delay_seconds,
+                    ),
+                )
+                self._retry(countdown=retry_delay_seconds)
```

Comment:
I’m not a big fan of putting more methods onto the "celery task", but instead move more code out of the "celery as task scheduling framework".

I think the only reason for this to be tied to "celery as a framework" is converting the exception to a retry.